### PR TITLE
Specify CompleteTrialSessions/Add more catches on exceptions/Update Readme

### PR DESCRIPTION
**Pull request by shenshan**

### CODE REVIEW

File: ibl_pipeline/analyses/behavior.py [hunk 1]
```diff
@@ -6,6 +6,69 @@
...
+    n_trials_stim_left=null:        blob    # number of trials when the stimulus is shown on the left
+    n_trials_stim_right=null:       blob    # number of trials when the stimulus is shown on the right
+    n_trials_stim_no_contrast=null: int     # number of trials when no stimulus is shown 
+    """
+
+    def make(self, key):
+        print(key)
+        key_con = key.copy()
+        trials = behavior.TrialSet.Trial & key
+        contrasts_left = np.unique(trials.fetch('trial_stim_contrast_left'))[1:] # discard contrast 0
```
Comment (dimitri-yatsenko) [inline ibl_pipeline/analyses/behavior.py]: If you fetch from the same expression, it's better to do it as one query to save network delays:
```python
contrast_left, contrast_right = trials.fetch('trial_stim_contrast_left', 'trial_stim_contrast_right')
contrast_left, contrast_right = np.unique(contrast_left), np.unique(contrast_right)
```

File: ibl_pipeline/analyses/behavior.py [hunk 2]
```diff
@@ -6,6 +6,69 @@
...
+        contrasts_right = np.unique(trials.fetch('trial_stim_contrast_right'))[1:] # discard contrast 0
```
Comment (dimitri-yatsenko) [inline ibl_pipeline/analyses/behavior.py]: Discarding the first element is not the same as "discard contrast 0". Is there a better way to exclude contrast 0?

File: ibl_pipeline/analyses/behavior.py [hunk 3]
```diff
@@ -6,6 +6,69 @@
...
+    contrasts_right=null:           blob    # contrasts shown on the right
+    n_trials_stim:                  blob    # number of trials for each contrast
+    n_trials_stim_left=null:        blob    # number of trials when the stimulus is shown on the left
+    n_trials_stim_right=null:       blob    # number of trials when the stimulus is shown on the right
+    n_trials_stim_no_contrast=null: int     # number of trials when no stimulus is shown 
+    """
+
+    def make(self, key):
+        print(key)
+        key_con = key.copy()
```
Comment (dimitri-yatsenko) [inline ibl_pipeline/analyses/behavior.py]: Is there a better name than `key_con`.  It's not used as a key and `con` is not clear. Let's follow PEP8

File: ibl_pipeline/analyses/behavior.py [hunk 4]
```diff
@@ -6,6 +6,69 @@
...
+            n_trials_stim_left = [len(trials & 'trial_stim_contrast_left={}'.format(contrast)) \
+                                    for contrast in contrasts_left]
+            key_con['n_trials_stim_left'] = n_trials_stim_left
+        else:
+            n_trials_stim_left = []
+            
+        if contrasts_right.size != 0:
+            key_con['has_right'] = True
+            key_con['contrasts_right'] = contrasts_right
+            n_trials_stim_right = [len(trials & 'trial_stim_contrast_right={}'.format(contrast)) \
```
Comment (dimitri-yatsenko) [inline ibl_pipeline/analyses/behavior.py]: Why use string conditions instead of dict conditions? `trials & {'trial_stim_contrast_right': contrast}`

Comment (dimitri-yatsenko) [reply ibl_pipeline/analyses/behavior.py]: Also, this loop can be replaced with a single `dj.U` query:
```
n_trials_stim_right = dj.U('contrast_right').aggr(trials, n='count(*)').fetch('n')
```

File: ibl_pipeline/analyses/behavior.py [hunk 5]
```diff
@@ -44,38 +110,71 @@ def make(self, key):
...
+        
+        key_rt = key.copy()
+        trials = behavior.TrialSet.Trial & key
+        contrasts_left = np.unique(trials.fetch('trial_stim_contrast_left'))[1:] # discard contrast 0
+        contrasts_right = np.unique(trials.fetch('trial_stim_contrast_right'))[1:] # discard contrast 0
+        trials_no_contrast = trials & 'trial_stim_contrast_right=0' & 'trial_stim_contrast_left=0'
+
+        trials_rt = trials.proj(rt='trial_response_time-trial_stim_on_time')
+
+        rt_stim_left = [(trials_rt & (trials & 'trial_stim_contrast_left={}'.format(contrast))).fetch('rt').mean() \
```
Comment (dimitri-yatsenko) [inline ibl_pipeline/analyses/behavior.py]: replace with `aggr` query

File: ibl_pipeline/analyses/behavior.py [hunk 6]
```diff
@@ -1,81 +1,191 @@
...
+        if contrasts_right.size != 0:
+            key_con['has_right'] = True
+            key_con['contrasts_right'] = contrasts_right
+            n_trials_stim_right = [len(trials & 'ABS(trial_stim_contrast_right-{})<1e-10'.format(contrast)) \
+                                    for contrast in contrasts_right]
+            key_con['n_trials_stim_right'] = n_trials_stim_right
+        else:
+            n_trials_stim_right = []
+        
+        if len(trials_no_contrast) > 0:
```
Comment (dimitri-yatsenko) [inline ibl_pipeline/analyses/behavior.py]:
```suggestion
        if trials_no_contrast:
```

File: ibl_pipeline/analyses/behavior.py [hunk 7]
```diff
@@ -1,81 +1,191 @@
...
+    reaction_time:     blob   # reaction time for all contrasts
+    """
+
+    key_source = behavior.TrialSet & \
+        (behavior.CompleteTrialSession & 'stim_on_times_status != "Missing"') & \
+        PsychResults
+
+    def make(self, key):
+        
+        key_rt = key.copy()
```
Comment (dimitri-yatsenko) [inline ibl_pipeline/analyses/behavior.py]: please choose a better name than `key_rt`. Use full words. Use `key` only for primary keys.

File: ibl_pipeline/analyses/behavior.py [hunk 8]
```diff
@@ -1,81 +1,191 @@
...
+        trials = behavior.TrialSet.Trial & key
+        contrasts_left = np.unique(trials.fetch('trial_stim_contrast_left'))[1:] # discard contrast 0
+        contrasts_right = np.unique(trials.fetch('trial_stim_contrast_right'))[1:] # discard contrast 0
+        trials_no_contrast = trials & 'trial_stim_contrast_right=0' & 'trial_stim_contrast_left=0'
+
+        trials_rt = trials.proj(rt='trial_response_time-trial_stim_on_time')
+
+        rt_stim_left = [(trials_rt & (trials & 'ABS(trial_stim_contrast_left-{})<1e-10'.format(contrast))).fetch('rt').mean() \
```
Comment (dimitri-yatsenko) [inline ibl_pipeline/analyses/behavior.py]: This can be performed as an `aggr`

File: ibl_pipeline/behavior.py [hunk 9]
```diff
@@ -343,20 +343,47 @@ class CompleteTrialSession(dj.Computed):
...
+                    stimOn_times = ONE().load(eID, dataset_types='_ibl_trials.stimOn_times')
+                
+                if np.all(np.isnan(stimOn_times)):
+                    key['stim_on_times_status'] = 'Missing'
+                elif np.any(np.isnan(stimOn_times)):
+                    key['stim_on_times_status'] = 'Partial'
+                else:
+                    key['stim_on_times_status'] = 'Complete'
+            
+            if '_ibl_trials.repNum.npy' not in datasets:
```
Comment (dimitri-yatsenko) [inline ibl_pipeline/behavior.py]:
```python
key['rep_num_status'] = 'Complete'  if '_ibl_trials.repNum.npy' in datasets else 'Missing'
```
or
```python
key['rep_num_status'] = ('Missing', 'Complete')['_ibl_trials.repNum.npy' in datasets]
```

File: prelim_analyses/behavioral_snapshots/behavioral_overview_perlab.py [hunk 10]
```diff
@@ -120,8 +118,24 @@
 			except:
 				pass
 
-			now = datetime.datetime.now().strftime("%Y-%m-%d")
-			fig.savefig(os.path.join(path + '%s_%s_batch_%s_%s.pdf'%(now, lab, birth_date, str(int(sub_batch/sub_batch_size)+1))))
-			fig.savefig(os.path.join(path + '%s_%s_batch_%s_%s.png'%(now, lab, birth_date, str(int(sub_batch/sub_batch_size)+1))))
+			mice_sub = mice[sub_batch:sub_batch+sub_batch_size]
+			mice_str = '"' + mice[0] + '"'
```
Comment (dimitri-yatsenko) [inline prelim_analyses/behavioral_snapshots/behavioral_overview_perlab.py]:
```python
mice_str = ', '.join('"%s"' % m for m in mice_sub)
```

File: ibl_pipeline/behavior.py [hunk 11]
```diff
@@ -343,20 +343,47 @@ class CompleteTrialSession(dj.Computed):
...
                             "_ibl_trials.contrastLeft.npy", "_ibl_trials.contrastRight.npy", \
-                            "_ibl_trials.stimOn_times.npy", "_ibl_trials.included.npy",\
                             "_ibl_trials.probabilityLeft.npy"]
 
     def make(self, key):
         datasets = (data.FileRecord & key & 'repo_name LIKE "flatiron_%"' & {'exists': 1}).fetch('dataset_name')
-        key['trial_session_complete'] = bool(np.all([req_ds in datasets for req_ds in self.required_datasets]))
-        if key['trial_session_complete'] is True:
+        is_complete = bool(np.all([req_ds in datasets for req_ds in self.required_datasets]))
+        if is_complete is True:
```
Comment (eywalker) [inline ibl_pipeline/behavior.py]: As you ensure `is_complete` is boolean above, `if is_complete is True:` is redundant and can simply be `if is_complete:`

**Pull request merged**